From 36743df4e87cfd4ad373d4858a49c4f50534db01 Mon Sep 17 00:00:00 2001
From: Collin Funk <collin.funk1@gmail.com>
Date: Mon, 12 Jan 2026 20:42:21 -0800
Subject: [PATCH] BACKPORT: maint: fix compilation errors with GCC 15

GCC 15 changed the default to -std=gnu23 where the current code would
cause -Wincompatible-pointer-types errors.

Reported by Ondrej Pohorelsky <opohorel@redhat.com> in
<https://lists.gnu.org/r/bug-time/2025-01/msg00000.html>
and by Khem Raj <raj.khem@gmail.com> in
<https://lists.gnu.org/r/bug-time/2025-03/msg00000.html>
and by Marcin Serwin <marcin@serwin.dev> in
<https://lists.gnu.org/r/bug-time/2025-10/msg00000.html>
and by Sam James <sam@gentoo.org> in
<https://savannah.gnu.org/bugs/?66450>.

* configure.ac (AC_TYPE_SIGNAL): Remove obsolete macro.
* src/time.c (sighandler): Remove type.
(run_command): Use sighandler_t which is defined by glibc or provided by
Gnulib.

(cherry picked from commit 4bfc492fcfa7778df7fc473ee901c1bdfdb02de0)
Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 configure.ac | 1 -
 src/time.c   | 6 +-----
 2 files changed, 1 insertion(+), 6 deletions(-)

diff --git a/configure.ac b/configure.ac
index ede8fd5..beb2077 100644
--- a/configure.ac
+++ b/configure.ac
@@ -66,7 +66,6 @@ dnl Checks for typedefs, structures, and compiler characteristics.
 AC_C_CONST
 AC_TYPE_PID_T
 AC_TYPE_SIZE_T
-AC_TYPE_SIGNAL
 
 dnl Checks for library functions.
 AC_FUNC_VPRINTF
diff --git a/src/time.c b/src/time.c
index 7e07995..68aadc6 100644
--- a/src/time.c
+++ b/src/time.c
@@ -74,10 +74,6 @@ enum
     "David MacKenzie",              \
     "Assaf Gordon"
 
-
-/* A Pointer to a signal handler.  */
-typedef RETSIGTYPE (*sighandler) ();
-
 /* msec = milliseconds = 1/1,000 (1*10e-3) second.
    usec = microseconds = 1/1,000,000 (1*10e-6) second.  */
 
@@ -728,7 +724,7 @@ run_command (cmd, resp)
      RESUSE *resp;
 {
   pid_t pid;			/* Pid of child.  */
-  sighandler interrupt_signal, quit_signal;
+  sighandler_t interrupt_signal, quit_signal;
   int saved_errno;
 
   resuse_start (resp);
-- 
2.53.0

